File: contrib/scaffold.go
package main

import (
	"flag"
	"fmt"
	"log"
	"net/url"
	"os"
	"path/filepath"
	"strings"
)

func main() {
	titlePtr := flag.String("title", "", "Title of a LeetCode problem (e.g., '1. Two Sum')")
	urlPtr := flag.String("url", "", "URL of the LeetCode problem")
	flag.Parse()

	if *titlePtr == "" {
		log.Fatal("-title is required")
	}
	if *urlPtr == "" {
		log.Fatal("-url is required")
	}

	pkgName := craftPkgName(*titlePtr)
	dirPath := filepath.Join("go", pkgName)

	if err := os.MkdirAll(dirPath, 0755); err != nil {
		log.Fatalf("Failed to create directory: %v", err)
	}

	if _, err := url.Parse(*urlPtr); err != nil {
		log.Fatal("URL is invalid:", err)
	}
	notePath := filepath.Join(dirPath, "note.md")
	if _, err := os.Stat(notePath); os.IsNotExist(err) {
		if err := os.WriteFile(notePath, []byte(craftNoteMd(*titlePtr, *urlPtr)+"\n"), 0644); err != nil {
			log.Fatalf("Failed to write note.md: %v", err)
		}
	}

	solutionPath := filepath.Join(dirPath, "solution.go")
	if _, err := os.Stat(solutionPath); os.IsNotExist(err) {
		if err := os.WriteFile(solutionPath, fmt.Appendf(nil, "package %s\n\nfunc __solution__() {\n}\n", pkgName), 0644); err != nil {
			log.Fatalf("Failed to write solution.go: %v", err)
		}
	}

	testPath := filepath.Join(dirPath, "solution_test.go")
	if _, err := os.Stat(testPath); os.IsNotExist(err) {
		if err := os.WriteFile(testPath, fmt.Appendf(nil, "package %s\n\nimport (\n\t\"testing\"\n)\n\nfunc Test__solution__(t *testing.T) {\n}\n", pkgName), 0644); err != nil {
			log.Fatalf("Failed to write solution_test.go: %v", err)
		}
	}

	fmt.Printf("Created: %s/\n", dirPath)
}

func craftNoteMd(title string, urlStr string) string {
	return fmt.Sprintf("[%s](%s)", title, urlStr)
}

func craftPkgName(title string) string {
	num, titleNoNum, _ := strings.Cut(title, ".")
	num = fmt.Sprintf("%04s", num)
	titleNoNumNormalized := strings.ToLower(strings.ReplaceAll(strings.TrimSpace(titleNoNum), " ", ""))
	return "s" + num + titleNoNumNormalized
}

File: contrib/Scaffold.java
void main(String[] args) throws Exception {
  var scriptPath = Path.of(getClass().getProtectionDomain().getCodeSource().getLocation().getPath());

  String title = "";
  String url = "";

  var i = 0;
  while (i < args.length) {
    var a = args[i];
    switch (a) {
      case "--title":
        title = args[i + 1];
        i += 2;
        break;
      case "--url":
        url = args[i + 1];
        i += 2;
        break;
      case "-h":
      case "--help":
        usage();
        i++;
        break;
      default:
        if (a.startsWith("-")) {
          System.err.println("Unknown option [ " + a + " ]");
          System.exit(3);
        }
        i++;
        break;
    }
  }

  if (title.isEmpty()) {
    System.err.println("--title is required");
    System.exit(1);
  }
  if (url.isEmpty()) {
    System.err.println("--url is required");
    System.exit(1);
  }

  if (!url.startsWith("http")) {
    System.err.println("URL is invalid");
    System.exit(1);
  }

  var pkgName = getPackageName(title);
  var srcPath = scriptPath.getParent().getParent().resolve("java/src/main/java/com/github/fobshippingpoint").resolve(pkgName);
  var testPath = scriptPath.getParent().getParent().resolve("java/src/test/java/com/github/fobshippingpoint").resolve(pkgName);
  Files.createDirectories(srcPath);
  Files.createDirectories(testPath);

  var notePath = srcPath.resolve("note.md");
  if (!Files.exists(notePath)) {
    Files.writeString(notePath, "[" + title + "](" + url + ")\n");
  }

  var solutionPath = srcPath.resolve("Solution.java");
  if (!Files.exists(solutionPath)) {
    Files.writeString(solutionPath, "package com.github.fobshippingpoint." + pkgName + ";" + """


    public class Solution {

    }
    """);
  }

  var testJavaPath = testPath.resolve("SolutionTest.java");
  if (!Files.exists(testJavaPath)) {
    Files.writeString(testJavaPath, "package com.github.fobshippingpoint." + pkgName + ";" + """


    import org.junit.jupiter.params.ParameterizedTest;
    import org.junit.jupiter.params.provider.Arguments;
    import org.junit.jupiter.params.provider.MethodSource;

    import java.util.stream.Stream;

    import static org.junit.jupiter.api.Assertions.assertEquals;

    class SolutionTest {
        public static Stream<Arguments> provider() {
            return Stream.of(
                    // Arguments.of("Should be duplicates since the element 1 occurs at the indices 0 and 3.",
                    //         new int[]{1, 2, 3, 1}, true),
                    // Arguments.of("Should not be duplicates since All elements are distinct.",
                    //         new int[]{1, 2, 3, 4}, false),
                    // Arguments.of("Should be duplicates since The element 1 occurs at the indices 0 and 1.",
                    //         new int[]{1, 1, 1, 3, 3, 4, 3, 2, 4, 2}, true)
            );
        }

        @ParameterizedTest
        @MethodSource("provider")
        void test__solution__() {

        }

    }
    """);
  }
}

String getPackageName(String title) {
  var dotIdx = title.indexOf('.');
  var num = String.format("%04d", Integer.parseInt(title.substring(0, dotIdx).trim()));
  var titleNoNum = title.substring(dotIdx + 1).replaceAll("[^\\p{Alnum}]", "").toLowerCase();
  return "s" + num + titleNoNum;
}

void usage() {
  IO.println("""
  Usage: java Scaffold.java --title title --url url 

  Create scaffold for LeetCode problem

  Options:
    -h, --help     You're looking at it
    --title        Title of a LeetCode problem (e.g., `1. Two Sum`)
    --url          URL of the LeetCode problem (e.g., `https://leetcode.com/problems/two-sum/`)
  """);
  System.exit(0);
}
